EvidenceApi/V1/UseCase/CreateResidentReferenceIdUseCase.cs

```diff
@@ -0,0 +1,56 @@
+using System.Collections.Generic;
+using System.Linq;
+using EvidenceApi.V1.Domain;
+using EvidenceApi.V1.Gateways.Interfaces;
+using EvidenceApi.V1.Infrastructure.Interfaces;
+using EvidenceApi.V1.UseCase.Interfaces;
+
+namespace EvidenceApi.V1.UseCase
+{
+    public class CreateResidentReferenceIdUseCase : ICreateResidentReferenceIdUseCase
+    {
+        private readonly IEvidenceGateway _evidenceGateway;
+        private readonly IStringHasher _stringHasher;
+        private const int ResidentReferenceLength = 11;
+
+        public CreateResidentReferenceIdUseCase(IEvidenceGateway evidenceGateway, IStringHasher stringHasher)
+        {
+            _evidenceGateway = evidenceGateway;
+            _stringHasher = stringHasher;
+        }
+
+        public string Execute(Resident resident)
+        {
+            var residentId = resident.Id;
+            var evidenceRequestsForResident = _evidenceGateway.FindEvidenceRequestsByResidentId(residentId);
+            if (evidenceRequestsForResident.Count > 0)
+            {
+                return evidenceRequestsForResident.First().ResidentReferenceId;
+            }
+
+            var residentIdHash = _stringHasher.create(residentId.ToString());
+            var hashStartIndex = 0;
+            var residentReferenceId = trimHash(residentIdHash, hashStartIndex, ResidentReferenceLength);
+            var allResidentReferenceIds = _evidenceGateway.GetAll().Select(e => e.ResidentReferenceId).ToHashSet();
```

Comment:
Shouldn't this be part of the `ResidentReferenceIdExists` function?

Reply:
My thinking is that we retrieve all of the existing evidence requests with `_evidenceGateway.GetAll` just once, and create a `Set` with all of the resident reference IDs. Then we keep using that to see if the trimmed hash exists already